data_structures/reverse_a_linked_list.cpp

```diff
@@ -0,0 +1,183 @@
+ * completion of an iteration.
+
+ *[A graphic explanation and view of what's happening behind the
+ *scenes](https://drive.google.com/file/d/1pM5COF0wx-wermnNy_svtyZquaCUP2xS/view?usp=sharing)
+ */
+
+#include <cassert>   /// for assert
+#include <iostream>  /// for I/O operations
+#include <memory>    /// for dynamic memory
+#include <new>       /// for managing  dynamic storage
+
+/**
+ * @namespace data_structures
+ * @brief Data Structures algorithms
+ */
+namespace data_structures {
+/**
+ * @namespace linked_list
+ * @brief Functions for singly linked list algorithm
+ */
+namespace linked_list {
+/**
+ * A Node class containing a value and pointer to another link
+ */
+class Node {
+ public:
+    int32_t val;  /// value of the current link
+    Node *next;   /// pointer to the next value on the list
+};
+
+/**
+ * A list class containing a sequence of links
+ */
+class list {
+ private:
+    Node *head;  // link before the actual first element
+ public:
+    /**
+     * List constructor. Initializes the first link.
+     */
+    list() {
+        head = nullptr;  // Initialize the first link
+    }
+    bool isEmpty();
+    void insert(int32_t new_elem);
+    void reverseList();
+    void display();
+    int32_t top();
+    int32_t last();
+};
+
+/**
+ * @brief function checks if list is empty
+ * @returns true if list is empty
+ * @returns false if list is not empty
+ */
+bool list::isEmpty() { return head == nullptr; }
+
+/**
+ * @brief function adds new element to the end of the list
+ * @param new_elem to be added to the end of the list
+ */
+void list::insert(int32_t n) {
+    try {
+        Node *new_node = new Node();
+        Node *temp = nullptr;
+        new_node->val = n;
+        new_node->next = nullptr;
+        if (isEmpty()) {
+            head = new_node;
+        } else {
+            temp = head;
+            while (temp->next != nullptr) {
+                temp = temp->next;
+            }
+            temp->next = new_node;
+        }
+    } catch (std::bad_alloc &exception) {
+        std::cerr << "bad_alloc detected: " << exception.what() << "\n";
+    }
+}
+
+/**
+ * @brief function reverseList for reversing the  list
+ * @brief Using current,previous and next pointer.
+ * @returns 'void'
+ */
+void list::reverseList() {
+    Node *curr = head;
+    Node *prev = nullptr, *next_node = nullptr;
+    while (curr != nullptr) {
+        next_node = curr->next;
+        curr->next = prev;
+        prev = curr;
+        curr = next_node;
+    }
+    head = prev;
+}
+
+/**
+ * @brief function to find the top element of the list
+ * @returns 'int32_t n'
+ * @brief returns the first element in the list
+ */
+int32_t list::top() {
+    try {
+        if (!isEmpty()) {
+            return head->val;
+        }
+    } catch (const std::exception &e) {
+        std::cerr << "List is empty" << e.what() << '\n';
+    }
+}
+/**
+ *  @brief function to find the last element of the list
+ *  @returns 'int32_t t->val'
```

Comment:
What does it return? Please document what does it return, not the variable or the variable's type. **Let us know if you need any help with this.** 🙂